### ISSUE TITLE
Incorrect before_node in GROUP_COEFFICIENTS change

### ISSUE THREAD

**Issue by aelnaiem** [socraticorg/mathsteps#120]
@karuppiah7890 noticed that one of the steps prints out the incorrect math statement. The final result is correct, so it's even more curious how this is happening. Most likely due to how we clone the node.

```
before change: 2x + 3x = 35
change: SIMPLIFY_LEFT_SIDE
after change: 5x = 35
# of substeps: 2
Substeps : 
before change: 2x + 3x = 35
change: GROUP_COEFFICIENTS
after change: (2 + 3)x = 35
# of substeps: 0
before change: 2 + 3x = 35 *This is wrong and should say 2x + 3x = 35*
change: SIMPLIFY_ARITHMETIC
after change: 5x = 35
# of substeps: 0

before change: 5x = 35
change: DIVIDE_FROM_BOTH_SIDES
after change: (5x) / 5 = 35/5
# of substeps: 0
before change: 5x / 5 = 35/5
change: SIMPLIFY_LEFT_SIDE
after change: x = 35/5
# of substeps: 0
before change: x = 35/5
change: SIMPLIFY_RIGHT_SIDE
after change: x = 7
# of substeps: 0
```

**Comment by Flyr1Q** [comment]
@aelnaiem @karuppiah7890 It seems like there is a difference between node.toString() and util/print.js in this case.

`
# Everything is fine for parenthesis:
Node {
  implicit: true,
  op: '*',
  fn: 'multiply',
  args:  [
     Node { content: [Object], changeGroup: 1 },
     Node { name: 'x', changeGroup: undefined }
  ],
  changeGroup: 1
}
# node.toString(): (2 + 3) x
# util/print.js: (2 + 3)x

# Once we get rid of parenthesis:
Node {
  implicit: true,
  op: '*',
  fn: 'multiply',
  args: [
    Node {
       implicit: false,
       op: '+',
       fn: 'add',
       args: [Object],
       changeGroup: 1
     },
     Node { name: 'x', changeGroup: undefined } ],
  changeGroup: undefined
}
# node.toString(): (2 + 3) x
# util/print.js: 2 + 3x
`

Actually, I would be glad to fix this. The final result is correct, so I assume we only need to fix util/print.js. Particularly this lines of code:
`
# util > print.js
# 44 if (node.implicit) {
# 45  break;
# 46 }
# ...
# 72 str += printTreeTraversal(node.args[i]);
`

Do I have to check if node is implicit & args.length > 1 and then surround with parenthesis?

**Comment by Flyr1Q** [comment]
I think #126 relate to this issue

**Comment by evykassirer** [comment]
yes, this sounds like a reasonable solution to this problem. I'd love if you could update `print` to add in parens better!

I'll assign you to this and to 126

**Comment by aelnaiem** [comment]
I love when issues are related!

**Comment by evykassirer** [comment]
sadly not related, by @Flyr1Q will look into this still :)

**Comment by Flyr1Q** [comment]
@evykassirer 

```
const math = require('mathjs');
const print = require('./lib/util/print');
const flatten = require('./lib/util/flattenOperands');
const collectAndCombineSearch = require('./lib/simplifyExpression/collectAndCombineSearch')

const node = math.parse('2x + 4x + y');
const status = collectAndCombineSearch(flatten(node));

status.substeps.forEach(substep => {
  console.log("before change: " + print(substep.oldNode));  // e.g. before change: 2x + 3x = 35
  console.log("change: " + substep.changeType);                  // e.g. change: SIMPLIFY_LEFT_SIDE
  console.log("after change: " + print(substep.newNode));   // e.g. after change: 5x = 35
});
```

```
```

**Comment by evykassirer** [comment]
oldNode needs to be updated as it goes up the tree, but it should always be separate from newNode

where do you think it's being changed?

(we can also chat on https://gitter.im/mathsteps-chat/Lobby if that's easier for you than github, but github here is fine as well!)
